src/pyspiffe/workloadapi/default_x509_source.py

```diff
@@ -0,0 +1,129 @@
+
+            picker: Function to choose the X.509 SVID from the list returned by the Workload API.
+                    If it is not set, the default SVID is picked. If the picker function throws an error,
+                    it will render the X509Source invalid and it will be closed.
+
+        Returns:
+            DefaultX509Source: New DefaultX509Source object, initialized with the X509Context fetched from the Workload API.
+
+        Raises:
+            ArgumentError: If spiffe_socket_path is invalid or not provided and SPIFFE_ENDPOINT_SOCKET env variable is not set.
+
+            X509SourceError: In case a timeout was configured and it was reached during the source initialization waiting
+                             for the first update from the Workload API.
+        """
+
+        self._initialized = threading.Event()
+        self._lock = threading.Lock()
+        self._closed = False
+        self._workload_api_client = (
+            workload_api_client
+            if workload_api_client
+            else DefaultWorkloadApiClient(spiffe_socket)
+        )
+        self._picker = picker
+
+        # set the watcher that will keep the source updated and log the underlying errors
+        self._client_cancel_handler = self._workload_api_client.watch_x509_context(
+            self._set_context, self._log_error
+        )
+
+        self._initialized.wait(timeout_in_seconds)
+
+        if not self._initialized.is_set():
+            self._client_cancel_handler.cancel()
+            raise X509SourceError(
+                'Could not initialize X.509 Source: reached timeout waiting for the first update'
+            )
+
+    def get_x509_svid(self) -> X509Svid:
+        """Returns an X509-SVID from the source."""
+        with self._lock:
+            if self._closed:
+                raise X509SourceError('Cannot get X.509 SVID: source is closed')
+        return self._x509_svid
+
+    def get_bundle_for_trust_domain(
+        self, trust_domain: TrustDomain
+    ) -> Optional[X509Bundle]:
+        """Returns the X.509 bundle for the given trust domain."""
+        with self._lock:
+            if self._closed:
+                raise X509SourceError('Cannot get X.509 Bundle: source is closed')
+        return self._x509_bundle_set.get_x509_bundle_for_trust_domain(trust_domain)
+
+    def close(self) -> None:
+        """Closes this X509Source closing the underlying connection with the Workload API. Once the source is closed,
+        no methods can be called on it.
+        """
+        with self._lock:
+            # the cancel method throws a grpc exception, that can be discarded
+            with suppress(Exception):
+                self._client_cancel_handler.cancel()
+            # prevents blocking on the constructor
+            self._initialized.set()
+            self._closed = True
+
+    def _set_context(self, x509_context: X509Context) -> None:
+        try:
+            if self._picker:
```

Comment:
_x509_svid and _x509_bundle_set access should be thread safe, this access should be protected too.